expr/context.go

```diff
@@ -0,0 +1,52 @@
+package expr
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/zcode"
+)
+
+// Context is an interface to a scope and value allocator for expressions.
+// This allows external packages to implement efficint temporary buffering
+// of Zed values both for let-style temporary variables accessible via
+// Context.Scope() and for allocating results.
+type Context interface {
+	Scope() []zed.Value
+	//XXX there should be two NewValues: one when bytes is already inside
+	// of the context... another when you need to copy those bytes into
+	// this context.
+	zed.Allocator
+}
+
+type allocator struct{}
+
+var _ Context = (*allocator)(nil)
+
+func NewContext() *allocator {
+	return &allocator{}
+}
+
+func (*allocator) NewValue(typ zed.Type, bytes zcode.Bytes) *zed.Value {
+	return zed.NewValue(typ, bytes)
+}
+
+func (*allocator) CopyValue(val zed.Value) *zed.Value {
+	return zed.NewValue(val.Type, val.Bytes)
```

Comment:
If the CopyValue contract is that it returns a value that does not share val.Bytes, then we need an actual copy here.
```suggestion
	return val.Copy()
```

Reply:
same as above